src/blocks/VoiceApp.tsx

```diff
@@ -0,0 +1,335 @@
+import Block from './Block'
+import { BlockModel } from './types'
+import './BlockStyles.css'
+import { createContext, SetStateAction, useContext, useEffect, useRef, useState, Dispatch, ReactNode } from 'react';
+import Card from '@mui/material/Card';
+import Box from '@mui/material/Box';
+import CardContent from '@mui/material/CardContent';
+import MicIcon from '@mui/icons-material/Mic';
+import Fab from '@mui/material/Fab';
+import PlayCircleIcon from '@mui/icons-material/PlayCircle';
+/*
+
+  TYPES
+
+*/
+
+type AudioContextProps = {
+  isRecording: boolean;
+  setIsRecording: Dispatch<SetStateAction<boolean>>
+  mediaRecorder: React.MutableRefObject<MediaRecorder | null>
+}
+
+type AudioButtonProps = {
+  onSave: (url: string) => void;
+}
+
+type PostInFeedProps = {
+  url: string
+}
+
+type AudioProviderProps = {
+  children: ReactNode;
+}
+
+type StartProps = {
+  node: AudioNode
+  context: AudioContext
+  getPlayable: (node: AudioNode, context: AudioContext) => boolean
+}
+
+/*
+
+  VoiceApp DATA
+
+*/
+
+const defaultAudioContext = { isRecording: false, setIsRecording: () => { }, mediaRecorder: { current: null } }
+
+const audioContext = createContext<AudioContextProps>(defaultAudioContext);
+
+const { Provider: AudioProvider } = audioContext;
+
+const AudioCtx = ({ children }: AudioProviderProps) => {
+  const [isRecording, setIsRecording] = useState(false);
+  const mediaRecorder = useRef<MediaRecorder | null>(null)
+
+
+  return (
+    <AudioProvider value={{ isRecording, setIsRecording, mediaRecorder }}>
+      {children}
+    </AudioProvider>
+  )
+
+}
+
+const start = ({ node, context, getPlayable }: StartProps) => {
+  // Hi, I'm doing the drawing
+  let analyser = context.createAnalyser();
+  analyser.smoothingTimeConstant = .9;
+  node.connect(analyser);
+
+  // ...
+
+  analyser.fftSize = 2048;
+  let bufferLength = analyser.frequencyBinCount;
+  let timeDomainData = new Uint8Array(bufferLength);
+  let frequencyData = new Uint8Array(bufferLength);
+  analyser.getByteTimeDomainData(timeDomainData);
+  analyser.getByteFrequencyData(frequencyData);
+
+  // Get a canvas defined with ID "oscilloscope"
+  let canvas = document.getElementById("oscilloscope") as HTMLCanvasElement;
```

Comment:
what happens here when there are multiple VoiceNote miniapps in the feed that all have that id? can you guarantee that it's always going to be the one for this instantiation of the miniapp?

Reply:
HA, good point. Fixing now.

Reply:
fixed